src/pymor/algorithms/rand_la.py

```diff
@@ -2,20 +2,183 @@
 # Copyright pyMOR developers and contributors. All rights reserved.
 # License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
 
+from numbers import Number
 import numpy as np
 import scipy as sp
 from scipy.sparse.linalg import eigsh, LinearOperator
 from scipy.linalg import lu_factor, lu_solve
 from scipy.special import erfinv
 
 from pymor.algorithms.gram_schmidt import gram_schmidt
+from pymor.core.cache import CacheableObject, cached
 from pymor.core.defaults import defaults
-from pymor.operators.interface import Operator
 from pymor.core.logger import getLogger
+from pymor.operators.interface import Operator
 from pymor.operators.constructions import InverseOperator, IdentityOperator
+from pymor.tools.deprecated import Deprecated
+from pymor.tools.random import get_seed_seq, new_rng
+
+
+class RandomizedRangeFinder(CacheableObject):
+    def __init__(self, A, subspace_iterations=0, range_product=None, source_product=None, lambda_min=None,
+                 complex=False):
+        assert isinstance(A, Operator)
+        assert 0 <= subspace_iterations and isinstance(subspace_iterations, int)
+        if range_product is None:
+            range_product = IdentityOperator(A.range)
+        if source_product is None:
+            source_product = IdentityOperator(A.source)
+        assert isinstance(range_product, Operator)
+        assert isinstance(source_product, Operator)
+        assert lambda_min is None or isinstance(lambda_min, Number)
+        assert isinstance(complex, bool)
+
+        self.__auto_init(locals())
+        self._l = 0
+        self._Q = [self.A.range.empty()]
+        for _ in range(subspace_iterations):
+            self._Q.append(self.A.source.empty())
+            self._Q.append(self.A.range.empty())
+        self._Q = tuple(self._Q)
+        self.testvecs = self.A.source.empty()
+        self._basis_rng_real = new_rng(get_seed_seq().spawn(1)[0])
+        self._test_rng_real = new_rng(get_seed_seq().spawn(1)[0])
+        if complex:
+            self._basis_rng_imag = new_rng(get_seed_seq().spawn(1)[0])
+            self._test_rng_imag = new_rng(get_seed_seq().spawn(1)[0])
```

Comment:
Consider the following

```python
rrf1 = RandomizedRangeFinder(op)
rrf2 = RandomizedRangeFinder(op)
with new_rng(0):
    Q1 = rrf1.find_range(10)
    stuff_that_uses_rng()
    Q1 = rrf1.find_range(20)

with new_rng(0):
    Q2 = rrf2.find_range(20)
```
`Q1` will not be equal to `Q2`. I would prefer them to be always equal independent of the call history. For this, the class must have its own `rng` that is exclusive as far as I can see.

A similar argument holds for the test vectors. They should be completely independent of the sample vectors. If they would share an `rng`, one would get different results of `find_range` depending on the history. E.g.for
```python
Q1 = rrf1.find_range(tol=1)
Q1 = rrf1.find_range(tol=0.1)
Q2 = rrf2.find_range(tol=0.1)
``` 
`Q1` and `Q2` would not be the same.

If we only had one `rng` for each, then in the complex case again the history would matter, because the random vectors are not drawn interlacingly. Having dedicated `rng`s for real and imaginary part eliminates that. They could be omitted, if the imaginary part is always drawn but only added onto the test vector in the complex case. The current implementation felt a little "safer" to me...

Reply:
> Consider the following
> 
> ```python
> with new_rng(0):
>     rrf1 = RandomizedRangeFinder(op)
>     Q1 = rrf1.find_range(10)
>     stuff_that_uses_rng()
>     Q1 = rrf1.find_range(20)
> 
> with new_rng(0):
>     rrf2 = RandomizedRangeFinder(op)
>     Q2 = rrf2.find_range(20)
> ```
> 
> `Q1` will not be equal to `Q2`. I would prefer them to be always equal independent of the call history.

Can you elaborate why you want/need that? The new randomness approach assumes that this should rarely be desired, so I frown upon any case where independence of call order is wanted ... 

> For this, the class must have its own `rng` that is exclusive as far as I can see.

One could write something like:

```python
rrf_rng = new_rng(0)
with rrf_rng:
    rrf1 = RandomizedRangeFinder(op)
    Q1 = rrf1.find_range(10)
stuff_that_uses_rng()
with rrf_rng:
    Q1 = rrf1.find_range(20)
```
This is not quite convenient, but doable.

> A similar argument holds for the test vectors. They should be completely independent of the sample vectors. 

Why? It's not like there is a correlation. It's a random algorithm, so the result are random anyway.

If they would share an `rng`, one would get different results of `find_range` depending on the history. E.g.for
> 
> ```python
> Q1 = rrf1.find_range(tol=1)
> Q1 = rrf1.find_range(tol=0.1)
> Q2 = rrf2.find_range(tol=0.1)
> ```
> 
> `Q1` and `Q2` would not be the same.
> 

I would claim that that's ok. (Although I must admit that it feels a little bit strange.)

> If we only had one `rng` for each, then in the complex case again the history would matter, because the random vectors are not drawn interlacingly. Having dedicated `rngs` for real and imaginary part eliminates that. They could be omitted, if the imaginary part is always drawn but only added onto the test vector in the complex case. The current implementation felt a little "safer" to me...

Reply:
Well then, I guess sometimes it's good to let go. I shall embrace _true randomness_.

Reply:
@artpelling, I think this is a good opportunity to think twice what we really want. If you disagree with my position, then, maybe, it would be good to have further discussions. I do not want to undo #1736 in a year or so. Otherwise, if you feel convinces, then I would merge #1736 soon.

Reply:
I am absolutely happy with #1736 and could not disagree with your position because I do not have any logical argument against it.

> Well then, I guess sometimes it's good to let go. I shall embrace _true randomness_.

I was not being sarcastic here. My previous implementation with the rngs was aimed at avoiding unexpected behaviour for the user. However, 'unexpected' depends on the users' expectations and you have convinced me that acutally nothing unexpected is happening.

However, it is important to make this clear. My suggestion would be to have a tutorial on randomness and `rand_la` with a caveat section that showcases some cases like above in order to avoid false expectations.

Reply:
Sound good! Do you want to write it? :smiley:

Reply:
It should have seen that one coming, shouldn't I?